### PR TITLE
Rewrite hoisting to work on new structure

### PR DESCRIPTION
**Pull request by jbruestle**
Fix hoisting to work on 3-function representation.  Right now, we only basic hoisting (no-side-effect ops on values already in the init) since I wanted to focus on the infrastructure + downstream passes, will re-add Rob's more complex hoisting support along with further tests once the rest of the infrastructure is is.

Even basic hoisting broke assumptions made delloc injection + tile->pxa, which I updated as well (including rename of BufferPlacement to DeallocPlacment which is what it was really doing).  Short version for each:

1) For deallocs, we need to put the deallocs of buffers that escape init via a pack into fini, not init.
2) For tile->pxa, there were some assumptions regarding how rank 0 tensors were lowered, sometimes into rank-0 memrefs and sometimes into scalars.  Basically, constants were loaded into scalars, and then we specifically looked for constants.  This meant that context was need across the pack/unpack boundary (i.e. original type -> post conversion type mapping was not fully determined without looking at defining ops).  I cleaned this up and set a fixed protocol for 0-dim tensors across the pack/unpack boundary.

### CODE REVIEW

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 1]
```diff
@@ -1564,9 +1560,32 @@ struct PackOpConversion : public OpConversionPattern<stdx::PackOp> {
...
+    SmallVector<Value, 8> scalarizedOperands;
+    for (auto val : operands) {
+      // Handle cases requring load
+      if (auto memrefType = val.getType().dyn_cast<MemRefType>()) {
+        if (memrefType.getRank() == 0) {
+          auto loadOp =
+              rewriter.create<pxa::PxaLoadOp>(op.getLoc(), val, ValueRange({}));
+          scalarizedOperands.push_back(loadOp.getResult());
+          Type foo = loadOp.getResult().getType();
+          IVLOG(3, "Packing scalarized type " << debugString(foo));
```
Comment (flaub) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1577]: Is this debugging code that can be removed? If not, rename `foo`.

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 2]
```diff
@@ -1564,9 +1560,32 @@ struct PackOpConversion : public OpConversionPattern<stdx::PackOp> {
...
+          continue;
+        }
+      }
+      // Default case is a no-op
+      Type foo = val.getType();
+      IVLOG(3, "Packing type " << debugString(foo));
```
Comment (flaub) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc:1583]: Rename `foo` or drop.

File: pmlc/dialect/pxa/transforms/passes.td [hunk 3]
```diff
@@ -18,9 +18,9 @@ def AutoTileExample : FunctionPass<"pxa-autotile-example"> {
   let constructor = "pmlc::dialect::pxa::createAutoTileExamplePass()";
 }
 
-def BufferPlacement : FunctionPass<"pxa-buffer-placement"> {
-  let summary = "Optimizes placement of alloc and dealloc operations";
-  let constructor = "pmlc::dialect::pxa::createBufferPlacementPass()";
+def DeallocPlacement : Pass<"pxa-dealloc-placement", "mlir::ModuleOp"> {
+  let summary = "Adds deallocs for memref allocss";
```
Comment (flaub) [inline pmlc/dialect/pxa/transforms/passes.td:22]: typo: `allocss`

**Review by flaub — COMMENTED** [review]
Minor comments

File: pmlc/conversion/tile_to_pxa/tile_to_pxa.cc [hunk 4]
```diff
@@ -1573,19 +1572,15 @@ struct PackOpConversion : public OpConversionPattern<stdx::PackOp> {
           auto loadOp =
               rewriter.create<pxa::PxaLoadOp>(op.getLoc(), val, ValueRange({}));
           scalarizedOperands.push_back(loadOp.getResult());
-          Type foo = loadOp.getResult().getType();
-          IVLOG(3, "Packing scalarized type " << debugString(foo));
           continue;
         }
       }
       // Default case is a no-op
       Type foo = val.getType();
```
Comment (flaub) [inline pmlc/conversion/tile_to_pxa/tile_to_pxa.cc]: Drop this if unused

**Comment by abialas1** [comment]
LGTM

**Review by flaub — APPROVED** [review]

**Pull request merged**